repository_instrument.go

```diff
@@ -403,6 +429,74 @@ func (r *instrumentRepository) DeleteInstrument(ctx context.Context, id uuid.UUI
 	return nil
 }
 
+func (r *instrumentRepository) CreateFtpConfig(ctx context.Context, ftpConfig *FTPConfig) error {
+	query := fmt.Sprintf(`INSERT INTO %s.sk_instrument_ftp_config(instrument_id, username, password, remote_path, file_mask, result_remote_path, file_suffix, ftp_server_type)
+		VALUES(:instrument_id, :username, :password, :remote_path, :file_mask, :result_remote_path, :file_suffix, :ftp_server_type)`, r.dbSchema)
+
+	dao := convertFtpConfigToDao(ftpConfig)
+	_, err := r.db.NamedExecContext(ctx, query, dao)
+	if err != nil {
+		log.Error().Err(err).Msg(msgCreateFtpConfigFailed)
+		return ErrCreateFtpConfigFailed
+	}
+	return nil
+}
+
+func (r *instrumentRepository) GetFtpConfigByInstrumentId(ctx context.Context, instrumentId uuid.UUID) (*FTPConfig, error) {
+	query := fmt.Sprintf(`SELECT * FROM %s.sk_instrument_ftp_config WHERE instrument_id = $1`, r.dbSchema)
+
+	var dao ftpConfigDAO
+	err := r.db.QueryRowxContext(ctx, query, instrumentId).StructScan(&dao)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			log.Warn().Interface("instrumentId", instrumentId).Msg("ftp config queried bot not found")
+			return nil, nil
+		}
+		log.Error().Err(err).Msg(msgGetFtpConfigFailed)
+		return nil, ErrGetFtpConfigFailed
+	}
+
+	return convertFtpConfigDaoToFtpConfig(dao), nil
+}
+
+func (r *instrumentRepository) FtpConfigExists(ctx context.Context, instrumentId uuid.UUID) (bool, error) {
+	query := fmt.Sprintf(`SELECT EXISTS(SELECT 1 FROM %s.sk_instrument_ftp_config WHERE instrument_id = $1)`, r.dbSchema)
+
+	var ftpConfigExists bool
+	err := r.db.QueryRowxContext(ctx, query, instrumentId).Scan(&ftpConfigExists)
+	if err != nil {
+		log.Error().Err(err).Msg(msgFtpConfigExistsFailed)
+		return ftpConfigExists, ErrFtpConfigExistsFailed
+	}
+
+	return ftpConfigExists, nil
+}
+
+func (r *instrumentRepository) UpdateFtpConfig(ctx context.Context, ftpConfig *FTPConfig) error {
+	query := fmt.Sprintf(`UPDATE %s.sk_instrument_ftp_config SET username = :username, password = :password,
+        remote_path = :remote_path, file_mask = :file_mask, result_remote_path = :result_remote_path, file_suffix = :file_suffix,
+        ftp_server_type = :ftp_server_type WHERE instrument_id = :instrument_id`, r.dbSchema)
+
+	dao := convertFtpConfigToDao(ftpConfig)
+	_, err := r.db.NamedExecContext(ctx, query, dao)
+	if err != nil {
+		log.Error().Err(err).Msg(msgUpdateFtpConfigFailed)
+		return ErrUpdateFtpConfigFailed
+	}
+	return nil
+}
+
+func (r *instrumentRepository) DeleteFtpConfig(ctx context.Context, instrumentId uuid.UUID) error {
+	query := fmt.Sprintf(`DELETE FROM %s.sk_instrument_ftp_config WHERE instrument_id = $1;`, r.dbSchema)
```

Comment:
In general we do soft deletes so that (config) changes can be tracked. I would recommend adding a separate primary key ID for the ftp_config with only foreign key reference to instrument_id, and adding a deleted_at column, and in this method just set the deleted_at to timezone('utc', now()), and adding WHERE deleted_at is NULL to other queries.

Reply:
Changed to soft deletes.